adapters/repos/db/lsmkv/snapshot.go

```diff
@@ -0,0 +1,114 @@
+package lsmkv
+
+import (
+	"context"
+	"io/fs"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	"github.com/semi-technologies/weaviate/entities/cyclemanager"
+	"github.com/semi-technologies/weaviate/entities/storagestate"
+)
+
+// PauseCompaction waits for all ongoing compactions to finish,
+// then makes sure that no new compaction can be started.
+//
+// This is a preparatory stage for taking snapshots.
+//
+// A timeout should be specified for the input context as some
+// compactions are long-running, in which case it may be better
+// to fail the backup attempt and retry later, than to block
+// indefinitely.
+func (b *Bucket) PauseCompaction(ctx context.Context) error {
+	compactionHalted := make(chan struct{})
+
+	go func() {
+		b.disk.compactionCycle.Stop(ctx)
```

Comment:
correct me if I'm wrong here:
- Stop acquire lock and because of unbuffered channel waits for cycle to stop
- context timeout is reached for long running cycle, Start is called (in defer)
-  Start can not proceed without lock, so it waits for it until cycle is stopped
-  Start is executed, PauseCompaction returns

Does it mean that method "ignores" the timeout and waits for cycle to finish anyway?
Same applies for FlushMemtable

Reply:
If the context is cancelled, `CycleManager.Stop(context.Context)` will immediately return and release the lock, so that `Start` can immediately obtain the lock. 

I have just pushed up a test demonstrating this.